bee-node/src/inner.rs

```diff
@@ -0,0 +1,239 @@
+// Copyright 2020 IOTA Stiftung
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and limitations under the License.
+
+use bee_common::shutdown;
+use bee_common_ext::{
+    node::{Node, NodeBuilder},
+    worker::Worker,
+};
+use bee_storage::storage::Backend;
+
+use anymap::{any::Any as AnyMapAny, Map};
+use async_trait::async_trait;
+use futures::{channel::oneshot, future::Future};
+use log::info;
+use tokio::spawn;
+
+use std::{
+    any::{type_name, Any, TypeId},
+    collections::{HashMap, HashSet},
+    marker::PhantomData,
+    pin::Pin,
+    sync::Arc,
+};
+
+type WorkerStart<N> = dyn for<'a> FnOnce(&'a mut N) -> Pin<Box<dyn Future<Output = ()> + 'a>>;
+type WorkerStop<N> = dyn for<'a> FnOnce(&'a mut N) -> Pin<Box<dyn Future<Output = ()> + Send + 'a>> + Send;
+
+#[allow(clippy::type_complexity)]
+pub struct BeeNode<B> {
+    workers: Map<dyn AnyMapAny + Send + Sync>,
+    tasks: HashMap<
+        TypeId,
+        Vec<(
+            oneshot::Sender<()>,
+            // TODO Result ?
+            Box<dyn Future<Output = Result<(), tokio::task::JoinError>> + Send + Sync + Unpin>,
+        )>,
+    >,
+    resources: Map<dyn AnyMapAny + Send + Sync>,
+    worker_stops: HashMap<TypeId, Box<WorkerStop<Self>>>,
+    worker_order: Vec<TypeId>,
+    phantom: PhantomData<B>,
+}
+
+impl<B: Backend> BeeNode<B> {
+    fn add_worker<W: Worker<Self> + Send + Sync>(&mut self, worker: W) {
+        self.workers.insert(worker);
+    }
+
+    fn remove_worker<W: Worker<Self> + Send + Sync>(&mut self) -> W {
+        self.workers
+            .remove()
+            .unwrap_or_else(|| panic!("Failed to remove worker `{}`", type_name::<W>()))
+    }
+}
+
+#[async_trait]
+impl<B: Backend> Node for BeeNode<B> {
+    type Builder = BeeNodeBuilder<B>;
+    type Backend = B;
+
+    async fn stop(mut self) -> Result<(), shutdown::Error>
+    where
+        Self: Sized,
+    {
+        for id in self.worker_order.clone().into_iter().rev() {
+            self.worker_stops.remove(&id).unwrap()(&mut self).await;
```

Comment:
Do we really want to stop *before* awaiting the task ?

Reply:
I'm actually not sure. Worker stopping and task stopping are currently pretty independent. I'll put the worker stop last if you wish, it's probably neater.

Reply:
I'm thinking of the tangle worker that logs things so we probably can't stop and drop before